File: cpp/creational/object_pool.cpp
/*Object Pool Pattern.
 *
 * > Uses a set of initialized objects
 * > kept ready to use, in a "pool",
 * > rather than allocating and destroying
 * > them on demand.
 *
 * This sample uses the pattern on the following
 * context:
 *     - An application has a huge set of AI models
 *         that work together to provide a given
 *         functionality;
 *     - Each model is fairly fast to run an inference;
 *     - Loading the models on the other hand requires
 *         the application to initiate a connection with
 *         a remote server that provides the most up-to-date
 *         version of the model;
 *     - Once loaded, the models are used for small periods
 *         of time on each call;
 */
#include <algorithm>
#include <chrono>
#include <iostream>
#include <memory>
#include <thread>
#include <vector>
using namespace std;

// === Utilities ===
struct Runtime_error {
  string message;
};

/*
 * Model interface.
 * */
template <typename K, typename T> class AbstractModel {
public:
  virtual vector<K> predict(vector<T> x) = 0;
  virtual void reset() {};
};

/*
 * Model pool.
 * */
template <typename K, typename T> class ModelPool {
private:
  vector<shared_ptr<AbstractModel<K, T>>> free, in_use;

public:
  ModelPool() {}

  void add(shared_ptr<AbstractModel<K, T>> model) {
    this->free.push_back(model);
  }

  shared_ptr<AbstractModel<K, T>> get() {
    if (free.empty()) {
      throw Runtime_error{"All objects are in_use or pool is empty."};
    }

    auto model = this->free[this->free.size() - 1];
    this->free.pop_back();
    this->in_use.push_back(model);
    return model;
  }

  void release(shared_ptr<AbstractModel<K, T>> model) {
    bool released = false;

    for (int i = 0; i < this->in_use.size(); i++) {
      auto &obj = this->in_use[i];
      if (model == obj) {
        this->free.push_back(model);
        this->in_use.erase(this->in_use.begin() + i);
        released = true;
      }
    }

    if (!released) {
      throw Runtime_error{"Object not managed by pool."};
    }
  }

  int free_count() { return this->free.size(); }

  int in_use_count() { return this->in_use.size(); }

  void print() {
    cout << "Pool(free=" << this->free_count()
         << ", in_use=" << this->in_use_count() << ")\n";
  }
};

/*
 * Sample model.
 * */
class SampleModel : public AbstractModel<int, int> {
private:
  vector<int> shift;

public:
  SampleModel() {
    for (int i = 0; i < 10; i++) {
      this->shift.push_back(random() % 21 - 10);
    }

    // Simulate slow initialization
    this_thread::sleep_for(chrono::milliseconds(500));
  }

  vector<int> predict(vector<int> x) {
    vector<int> out;
    int size = shift.size(), i = 0;
    auto fn = [size, *this, &i, &out](int v) {
      out.push_back(v + this->shift[i++ % size]);
    };
    for_each(x.begin(), x.end(), fn);
    return out;
  }
};

/*
 * Sample execution.
 * */
int main() {
  // Initialize RNG
  srand(time(nullptr));

  // Initialize pool
  ModelPool<int, int> pool;
  cout << "Empty pool created: ";
  pool.print();

  // Add some models
  cout << "Creating some models...\n";
  for (int i = 0; i < 3; i++) {
    pool.add(make_shared<SampleModel>());
  }
  pool.print();

  // Get some models
  cout << "Locking models...\n";
  vector<shared_ptr<AbstractModel<int, int>>> models{pool.get(), pool.get(),
                                                     pool.get()};
  pool.print();

  // Utility vector print
  int counter = 0;
  auto print = [&counter](int value) {
    if (counter == 0) {
      cout << "[";
    }

    cout << value;

    if (1 + counter++ < 3) {
      cout << ", ";
    } else {
      counter = 0;
      cout << "]";
    }
  };

  // Run predictions
  cout << "Running predictions:\n";
  vector<int> x = {1, 2, 3};
  for (int i = 0; i < models.size(); i++) {
    vector<int> fx = models[i]->predict(x);
    pool.release(models[i]);

    cout << "Model " << i + 1 << "\n\tx=";
    for_each(x.begin(), x.end(), print);
    cout << "\n\tf(x)=";
    for_each(fx.begin(), fx.end(), print);
    cout << "\n\t";
    pool.print();

    if (i + 1 < models.size()) {
      cout << "\n";
    }
  }

  return 0;
}

File: python/creational/object_pool.py
"""Object Pool Pattern.

> Uses a set of initialized objects
> kept ready to use, in a "pool",
> rather than allocating and destroying
> them on demand.

This sample uses the pattern on the following
context:
    - An application has a huge set of AI models
        that work together to provide a given
        functionality;
    - Each model is fairly fast to run an inference;
    - Loading the models on the other hand requires
        the application to initiate a connection with
        a remote server that provides the most up-to-date
        version of the model;
    - Once loaded, the models are used for small periods
        of time on each call;
"""

import time
from typing import Generic, TypeVar
import random
from abc import ABC, abstractmethod

T = TypeVar("T")
K = TypeVar("K")


class AbstractModel(ABC, Generic[K, T]):
    """Abstract model interface.

    A model is something that allows
    running predictions.
    """

    @abstractmethod
    def predict(self, x: list[K]) -> list[T]:
        """Run an inference.

        :param x: list of inputs.
        :return: list of outputs.
        """

    @abstractmethod
    def reset(self):
        """Reset state of the model."""


class ModelPool(Generic[K, T]):
    """Model pool."""

    def __init__(self):
        self._objects: dict[str, dict] = dict(free=dict(), in_use=dict())

    def add(self, model: AbstractModel[K, T]):
        key = id(model)
        assert all(key not in self._objects[k] for k in ["free", "in_use"])
        self._objects["free"][key] = model

    def get(self) -> AbstractModel[K, T]:
        key, model = self._objects["free"].popitem()
        self._objects["in_use"][key] = model
        return model

    def release(self, model: AbstractModel[K, T]):
        key = id(model)
        assert key in self._objects["in_use"]
        del self._objects["in_use"][key]
        model.reset()
        self._objects["free"][key] = model

    def free_count(self) -> int:
        return len(self._objects["free"])

    def in_use_count(self) -> int:
        return len(self._objects["in_use"])

    def __str__(self) -> str:
        return f"Pool(free={self.free_count()}, in_use={self.in_use_count()})"


class SampleModel(AbstractModel[int, int]):
    def __init__(self):
        self._shift = [random.randint(-10, 10) for _ in range(10)]

        # Simulate slow initialization
        time.sleep(0.5)

    def predict(self, x: list[int]) -> list[int]:
        return [x_ + self._shift[i % len(self._shift)] for i, x_ in enumerate(x)]

    def reset(self):
        # No-op, stateless model
        pass


if __name__ == "__main__":
    # Initialize pool
    pool = ModelPool[int, int]()
    print(f"Empty pool created: {pool}")

    # Add some models to the pool
    print("Creating some models...")
    for _ in range(3):
        pool.add(SampleModel())

    # Show pool state
    print(pool)

    # Get some models
    print("Locking models...")
    model_1 = pool.get()
    model_2 = pool.get()
    model_3 = pool.get()
    print(pool)

    # Run predictions
    print("Running predictions:")
    for i, m in enumerate([model_1, model_2, model_3]):
        x = [1, 2, 3]
        print(f"Model {i+1}\n\tx={x}, f(x)={m.predict(x)}\n\tReleasing model.")
        pool.release(m)
        print(f"\t{pool}")
        if i + 1 < 3:
            print()
